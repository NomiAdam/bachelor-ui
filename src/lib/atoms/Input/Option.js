// eslint-disable-next-line no-unused-vars
import React from 'react';
import styled from 'styled-components';
import { lightTheme } from '../../constants/theme';

const StyledOption = styled.option`
 	position: absolute;
	padding: 20px 2%;
    width: 96%;
    display: block;
    margin: 0px;
    cursor: pointer;
    background-color: transparent;
    font-size: 16px;
    transition: .5s all ease;
    &:hover {
    	background-color: ${ lightTheme.whiteBlue };
    };
`;

export default StyledOption;
