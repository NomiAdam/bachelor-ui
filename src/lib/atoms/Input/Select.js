// eslint-disable-next-line no-unused-vars
import React from 'react';
import styled from 'styled-components';
import { basicTheme } from '../../constants/theme';

export const StyledSelect = styled.select`
	height: 30px;
  	min-width: 50px;
  	padding: .25em;
  	border: 0 solid transparent;
  	border-radius: 3px;
  	border-bottom: ${ ({ withBorder }) => (withBorder ? `2px solid ${ basicTheme.blue }` : '0px') };
  	overflow-y: scroll;
  	font-weight: bold;
  	letter-spacing: .15em;
  	${ ({ error }) => (error ? `
	    border: 1px solid ${ basicTheme.red };
	    border-top-left-radius: 5px;
        border-top-right-radius: 5px;
	` : '') }
  	&:focus {
  		outline: none;
  	}
  	box-shadow: 1px 2px 10px -2px ${ basicTheme.grey };
`;

export default StyledSelect;
