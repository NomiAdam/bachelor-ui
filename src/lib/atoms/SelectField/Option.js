import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { resolveTheme, TYPES } from '../../utils/resolveTheme';

const StyledOption = styled.div`
    width: 100%;
    display: inline-block;
    border: none;
	  padding: 10px 20px;
	  text-align: center;
	  background-color: ${ ({ selected, ...props }) => (selected ? resolveTheme(TYPES.BACKGROUND_COLOR)(props) : 'initial') };
    &:hover {
        background-color: ${ resolveTheme(TYPES.BACKGROUND_COLOR_HOVER) };
        cursor: pointer;
    }
`;

const Option = ({
    value, onClick, label, selected, secondary,
}) => (
    <StyledOption secondary={ secondary } onMouseDown={ onClick(value) } selected={ selected }>
        {label}
    </StyledOption>
);

Option.propTypes = {
    /**
     * String value of Option
     */
    value: PropTypes.any,
    /**
     * onMouseDown event handler, (value) => (e) =>
     */
    onClick: PropTypes.func.isRequired,
    /**
     * String representation of displayed value
     */
    label: PropTypes.string,
    /**
     * Boolean representation of currently selected item
     */
    selected: PropTypes.bool,
};

export { StyledOption };
export default Option;
