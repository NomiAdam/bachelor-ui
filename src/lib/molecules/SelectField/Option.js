import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { lightTheme } from '../../constants/theme';

const StyledOption = styled.div`
    width: 100%;
    display: inline-block;
    border: none;
	padding: 10px 20px;
	text-align: center;
	background-color: ${ ({ selected }) => (selected ? lightTheme.lightGrey : 'initial') };
    &:hover {
        background-color: ${ lightTheme.lightGrey };
        cursor: pointer;
    }
`;

const Option = ({
    value, onClick, label, selected,
}) => (
    <StyledOption onMouseDown={ onClick(value) } selected={ selected }>
        {label}
    </StyledOption>
);

Option.propTypes = {
    /**
     * String value of Option
     */
    value: PropTypes.any,
    /**
     * onMouseDown event handler, (value) => (e) =>
     */
    onClick: PropTypes.func.isRequired,
    /**
     * String representation of displayed value
     */
    label: PropTypes.string,
    /**
     * Boolean representation of currently selected item
     */
    selected: PropTypes.bool,
};

export { StyledOption };
export default Option;
