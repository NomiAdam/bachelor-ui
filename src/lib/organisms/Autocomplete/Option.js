import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {lightTheme} from '../../constants/theme';

const StyledOption = styled.div`
    width: 100%;
    border: none;
	padding: 10px 20px;
	background-color: ${({selected}) => selected ? lightTheme.lightGrey : 'initial'};
    &:hover {
        background-color: ${lightTheme.lightGrey};
        cursor: pointer;
    }
`;

const Option = ({value, onClick, selected}) => (
    <StyledOption onMouseDown={onClick(value)} selected={selected}>
        {value}
    </StyledOption>
);

Option.propTypes = {
    /**
     * String value of Option
     */
    value: PropTypes.string,
    /**
     * onMouseDown event handler, (value) => (e) =>
     */
    onClick: PropTypes.func.isRequired,
    /**
     * Boolean value whether current option is selected or not
     */
    selected: PropTypes.bool,
};

export {StyledOption};
export default Option;
