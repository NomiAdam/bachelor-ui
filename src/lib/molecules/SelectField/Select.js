import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
    map, length, equals, o,
} from 'ramda';
import Option from './Option';
import { darkTheme, lightTheme } from '../../constants/theme';
import DEFAULT_THEME from '../../utils/resolveTheme';

const StyledSelect = styled.div`
    ${ DEFAULT_THEME }
    overflow: hidden;
    display: ${ props => (props.visible ? 'flex' : 'none') }
    top: -50px;
    position: absolute;
    z-index: 999;
    width: ${ ({ width }) => width };
    min-height: 0px;
    max-height: 500px;
    justify-content: center;
    align-items: center;
    flex-direction: column;
    ${ ({ isEmpty }) => (isEmpty ? 'border: none;' : '') }
    font-size: 16px;
  	box-shadow: 5px 5px 15px ${ darkTheme.darkGrey };
`;

const doesItHaveOptions = o(equals(0), length);
const Select = ({
    visible, options, onClick, width, selected, backgroundColor,
}) => {
    const equalsSelected = equals(selected);
    const mapSuggestion = ({ value, label }) => (
        <Option value={ value } label={ label } onClick={ onClick } key={ value } selected={ equalsSelected(value) } />
    );
    const mappedOptions = map(mapSuggestion);
    return (
        <StyledSelect
            backgroundColor={ backgroundColor }
            visible={ visible }
            isEmpty={ doesItHaveOptions(options) }
            width={ width }
        >
            {visible && mappedOptions(options)}
        </StyledSelect>
    );
};

Select.propTypes = {
    /**
     * Whether select div is visible or not, is visible when Input is focused
     */
    visible: PropTypes.bool,
    /**
     * Array of possible values [ { value: 'Value', label: 'Label' }, ... ]
     */
    options: PropTypes.array,
    /**
     * Option onMouseDown handler
     */
    onClick: PropTypes.func,
    /**
     * String representation of select width
     */
    width: PropTypes.string,
    /**
     * String definition of currently selected item
     */
    selected: PropTypes.string,
    /**
   * Optional String color definition
   */
    backgroundColor: PropTypes.string,
};

Select.defaultProps = {
    options: [],
    backgroundColor: lightTheme.white,
};

export { StyledSelect };
export default Select;
