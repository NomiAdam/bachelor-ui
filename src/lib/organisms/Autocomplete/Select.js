import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
    map, length, equals, o,
} from 'ramda';
import Option from './Option';
import { lightTheme } from '../../constants/theme';

const StyledSelect = styled.div`
    display: ${ props => (props.visible ? 'block' : 'none') }
    position: absolute;
    z-index: 999;
    background-color: ${ lightTheme.white };
    width: 100%;
    min-height: 0px;
    max-height: 500px;
    border: 1px solid ${ lightTheme.lightGrey }; 
    border-top: none;
    border-bottom-right-radius: 4px;
    border-bottom-left-radius: 4px;
    ${ props => (props.isEmpty ? 'border: none;' : '') }
    font-weight: 300;
    font-size: 16px;
    overflow-y: auto;
`;

const doesItHaveOptions = o(equals(0), length);
const Select = ({
    visible, suggestions, onClick, selected,
}) => {
    const equalsSelected = equals(selected);
    const mapSuggestion = ({ value }) => (
        <Option value={ value } onClick={ onClick } key={ value } selected={ equalsSelected(value) } />
    );
    const mappedOptions = map(mapSuggestion);
    return (
        <StyledSelect visible={ visible } isEmpty={ doesItHaveOptions(suggestions) }>
            {visible && mappedOptions(suggestions)}
        </StyledSelect>
    );
};

Select.propTypes = {
    /**
     * Whether select div is visible or not, is visible when Input is focused
     */
    visible: PropTypes.bool,
    /**
     * Array of possible values [ { value: 'Value', label: 'Label' }, ... ]
     */
    suggestions: PropTypes.array,
    /**
     * Option onMouseDown handler
     */
    onClick: PropTypes.func,
    /**
     * String definition of selected suggestion value
     */
    selected: PropTypes.string,
};

Select.defaultProps = {
    suggestions: [],
};

export { StyledSelect };
export default Select;
