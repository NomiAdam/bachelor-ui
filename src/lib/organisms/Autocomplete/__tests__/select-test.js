/* eslint-disable no-undef */
import React from 'react';
import Select, {StyledSelect} from '../Select';
import Option from '../Option';
import {shallow} from 'enzyme';
import 'jest-styled-components';

const mockClick = jest.fn();
const mockSuggestions = [{value: 'Value', label: 'Label'}];

describe('AutocompleteSelect', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });
    it('should be defined', () => {
        expect(Select).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<Select />)).toMatchSnapshot();
    });
    it('should match the snapshot with suggestions', () => {
        expect(shallow(<Select visible={true} suggestions={mockSuggestions} onClick={mockClick}/>)).toMatchSnapshot();
    });
    it('should be of type StyledSelect', () => {
        expect(shallow(<Select />).type()).toBe(StyledSelect);
    });
    it('should have matching length when visible of Option components', () => {
        const wrapper = shallow(<Select visible={true} suggestions={mockSuggestions} onClick={mockClick} />);
        expect( wrapper.find(Option).length).toBe(mockSuggestions.length);
    });
    it('should have length of zero when not visible', () => {
        const wrapper = shallow(<Select visible={false} suggestions={mockSuggestions} />);
        expect( wrapper.find(Option).length).toBe(0);
    });
    it('should not fire onClick action on render', () => {
        shallow(<Select onClick={mockClick}/>);
        expect(mockClick).not.toHaveBeenCalled();
    });
    it('should have border set to none when suggestions are empty Array', () => {
        expect(shallow(<Select suggestions={[]} />)).toHaveStyleRule('border','none');
    });
});

