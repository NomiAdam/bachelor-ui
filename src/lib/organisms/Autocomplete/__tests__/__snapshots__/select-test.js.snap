// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteSelect should match the snapshot 1`] = `
<styled.div
  isEmpty={true}
/>
`;

exports[`AutocompleteSelect should match the snapshot with suggestions 1`] = `
<styled.div
  isEmpty={false}
  visible={true}
>
  <Option
    key="Value"
    onClick={[MockFunction]}
    selected={false}
    value="Value"
  />
</styled.div>
`;
